server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSDelayOptionsCheckUtility.java

```diff
@@ -0,0 +1,86 @@
+ *
+ */
+package org.eclipse.lsp.cobol.implicitDialects.cics.utility;
+
+        import org.antlr.v4.runtime.ParserRuleContext;
+
+        import org.eclipse.lsp.cobol.common.dialects.DialectProcessingContext;
+        import org.eclipse.lsp.cobol.common.error.ErrorSeverity;
+        import org.eclipse.lsp.cobol.common.error.SyntaxError;
+        import org.eclipse.lsp.cobol.implicitDialects.cics.CICSLexer;
+        import org.eclipse.lsp.cobol.implicitDialects.cics.CICSParser;
+
+        import java.util.HashMap;
+        import java.util.List;
+        import java.util.Map;
+
+        import static org.eclipse.lsp.cobol.implicitDialects.cics.CICSParser.*;
+
+/** Checks CICS Delay rules for required and invalid options */
+public class CICSDelayOptionsCheckUtility extends CICSOptionsCheckBaseUtility {
+
+    public static final int RULE_INDEX = RULE_cics_delay;
+
+    private static final Map<Integer, ErrorSeverity> DUPLICATE_CHECK_OPTIONS =
+            new HashMap<Integer, ErrorSeverity>() {
+                {
+                    put(CICSLexer.DELAY, ErrorSeverity.ERROR);
+                    put(CICSLexer.INTERVAL, ErrorSeverity.ERROR);
+                    put(CICSLexer.TIME, ErrorSeverity.ERROR);
+                    put(CICSLexer.FOR, ErrorSeverity.ERROR);
+                    put(CICSLexer.HOURS, ErrorSeverity.ERROR);
+                    put(CICSLexer.MINUTES, ErrorSeverity.ERROR);
+                    put(CICSLexer.SECONDS, ErrorSeverity.ERROR);
+                    put(CICSLexer.MILLISECS, ErrorSeverity.ERROR);
+                    put(CICSLexer.UNTIL, ErrorSeverity.ERROR);
+                    put(CICSLexer.REQID, ErrorSeverity.ERROR);
+                }
+            };
+
+    public CICSDelayOptionsCheckUtility(
+            DialectProcessingContext context, List<SyntaxError> errors) {
+        super(context, errors, DUPLICATE_CHECK_OPTIONS);
+    }
+
+    /**
+     * Entrypoint to check CICS Delay rule options
+     *
+     * @param ctx ParserRuleContext subclass containing options
+     * @param <E> A subclass of ParserRuleContext
+     */
+    public <E extends ParserRuleContext> void checkOptions(E ctx) {
+        if (ctx.getRuleIndex() == RULE_cics_delay_opts) {
+            checkOpts((CICSParser.Cics_delay_optsContext) ctx);
+
+            checkDuplicates(ctx);
+        }
+    }
+
+    private void checkOpts(CICSParser.Cics_delay_optsContext ctx) {
+
+        if (!ctx.UNTIL().isEmpty())
+            checkHasIllegalOptions(ctx.MILLISECS(), "MILLISECS");
+
+        checkHasMutuallyExclusiveOptions("INTERVAL, TIME, UNTIL, FOR", ctx.INTERVAL(),
+                    ctx.TIME(), ctx.UNTIL(), ctx.FOR());
+
+        if (!ctx.INTERVAL().isEmpty() || !ctx.TIME().isEmpty()) {
+            checkHasIllegalOptions(ctx.HOURS(), "HOURS");
+            checkHasIllegalOptions(ctx.MINUTES(), "MINUTES");
+            checkHasIllegalOptions(ctx.SECONDS(), "SECONDS");
+            checkHasIllegalOptions(ctx.MILLISECS(), "MILLISECS");
+        }
```

Comment:
"DELAY FOR" and "DELAY UNTIL" should give an error message
`'TIME UNIT(s)' OPTION REQUIRED BUT NOT SPECIFIED.`